src/prng/mod.rs

```diff
@@ -34,20 +34,20 @@
 //!
 //! # The generators
 //!
-//! ## Basic pseudo-random number generators (PRNGs)
+//! ## Normal pseudo-random number generators (PRNGs)
 //!
-//! The goal of non-cryptographic PRNGs is usually to find a good balance between
-//! simplicity, quality, memory usage and performance. These algorithms are
-//! very important to Monte Carlo simulations, and also suitable for several
-//! other problems such as randomised algorithms and games (except where there
-//! is a risk of players predicting the next output value from previous values,
-//! in which case a CSPRNG should be used).
+//! The goal of normal, non-cryptographic PRNGs is usually to find a good
+//! balance between simplicity, quality, memory usage and performance. These
+//! algorithms are very important to Monte Carlo simulations, and also suitable
+//! for several other problems such as randomized algorithms and games (except
+//! where there is a risk of players predicting the next output value from
+//! previous values, in which case a CSPRNG should be used).
 //!
 //! Currently Rand provides only one PRNG, and not a very good one at that:
 //!
-//! | name | performance | worst-case | memory usage | initialisation | quality | predictability |
-//! |----- |------------ |----------- | -------------|--------------- |-------- |--------------- |
-//! | [`XorShiftRng`] | fast | fast | 16 bytes | fast | poor | trivial after 4 words |
+//! | name | full name | performance | memory | quality | period | features |
+//! |------|-----------|-------------|--------|---------|--------|----------|
+//! | [`XorShiftRng`] | Xorshift 32/128 | ⭐⭐⭐ | 16 bytes | ⭐ | `u32` * 2<sup>128</sup> - 1 | — |
```

Comment:
Good job on the stars.

But why does Xorshift only get 3 stars for performance, less than most of the crypto generators? It's much faster than ChaCha and I don't want to see distinct performance ratings.

Personally I think "predictability" is important; it has implications on how suitable the generator is for seeding other generators and theoretically influences quality.

Reply:
Predictability is problematic, because it is closely related to the cryptographic definition of randomness, which is binary: an RNG is either distinguishable from randomness or it is not. It should be equivalent to the distinction between CSPRNG and non-CSPRNG, anything else would be misleading IMHO.

Reply:
> why does Xorshift only get 3 stars for performance, less than most of the crypto generators?

I am not sure yet how to rate the performance. Do we want it to be comparable between both tables? Should there be ratings depending on the architecture?

I had having in mind something like:
- 5 stars: SIMD or something like that.
- 4 stars: uses a handful of 64-bit operations.
- 3 starts: uses a handful of 32-bit operations.
- 2 and 1 stars: uses many instructions, can not use instruction-level parallelism, combines multiple RNGs.

And for CSPRNGs:
- 5 stars: HC-128 and ISAAC-64, as I don't expect to see something significantly faster. Or maybe it should be 4, with hardware-accelerated AES as 5 stars.
- 3 stars: ChaCha20 somewhere in the middle.

I'm not attached to this, open for discussion :smile:.

Reply:
> Personally I think "predictability" is important; it has implications on how suitable the generator is for seeding other generators and theoretically influences quality.

I believe it works the other way around. The operation that removes practical correlations is also what makes it a bit harder to predict.

> Predictability is problematic, because it is closely related to the cryptographic definition of randomness, which is binary: an RNG is either distinguishable from randomness or it is not. It should be equivalent to the distinction between CSPRNG and non-CSPRNG, anything else would be misleading IMHO.

@vks We agree on something :smile:. We can note many things in the table, but I would be careful about predictability. It seems like something that can easily give the wrong impression.

Reply:
What do you think about stars based on the performance of `next_u64` on a current i5/i7?
- 5 stars 8000 MB/s+
- 4 stars 4000 MB/s+
- 3 stars 2000 MB/s+
- 2 stars 1000 MB/s+
- 1 star < 1000 MB/s

Reply:
> which is binary: an RNG is either distinguishable from randomness or it is not

What world of binaries do you come from? The world I live in has many unknowns. Cryptographic security isn't the only point either. But as I said in the text (which I think @pitdicker removed), non-crypto PRNGs can be categorised as "trivially predictable" or "possibly predictable". I still think it's worth pointing out that some PRNGs as *trivially* predictable while others are not quite so easy.

Benchmarks might be good for performance, but in that case we should probably define some CPU/system we use, and might as well just post the numbers. And then there are big differences between byte-array and `u64` outputs. So not sure.

Reply:
> What world of binaries do you come from? The world I live in has many unknowns.

I was not saying anything about "the world", so I really don't understand what your point is. My point was that there is no such thing as a "slightly  cryptographically secure PRNG" and I think "predictable with effort" suggests there is, so I don't think it's a good idea to make such claims.

From a practical perspective, I think you either want an unpredictable PRNG, which rules out any non-CSPRNG, or you don't care about predictability. You are right that being unpredictable is indicative of quality, but I think this should be absorbed in the quality rating.

In summary, I think making weakly founded claims about predictability is potentially misleading while offering little benefit, so I would prefer not to make such claims.

About the benchmarks I agree with you, instead of stars we might just use the actual numbers. Maybe use cycles / byte, this should give nice values close to 1.

Reply:
Fine, remove the "predictability" stuff for basic PRNGs then. You're right about potential misunderstanding.

Cycles / bytes sounds good in theory but (a) is still highly dependent on CPU type, (b) is less easy to interpret for users and (c) unfortunately there is a significant difference in performance depending on which of next_u32/next_u64/fill_bytes is used.